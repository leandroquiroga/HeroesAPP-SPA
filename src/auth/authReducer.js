import { types } from "../types/type";



export const authReducer = (state = {}, action) => {
  const { type, payload } = action;
  
  switch (type) {
    case types.login:
      return {
        ...payload,
        logged: true,
      };
    case types.logout:
      return {
        logged: false,
      };
    
    default:
      return state;
  }
}