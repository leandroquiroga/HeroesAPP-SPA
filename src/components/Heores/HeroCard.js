import React from 'react';
import { Link } from 'react-router-dom';

export const HeroCard = ({
  id,
  superhero,
  publisher, 
  alter_ego,
  first_appearance,
  characters,
}) => {
  const imageSRC = `/assets/img/${id}.jpg`
  const infoImagePath = `/assets/svg/info.svg`
  return (

    <div className='col animate__animated animate__fadeIn'>
      
      <div className='card p-3' style={{ width: 18 + 'rem'}}>  
        <img
          src={imageSRC}
          className='card-img-top img-thumbnail'
          alt={superhero}
        />

        <div className='card-body d-flex flex-column justify-content-center align-items-center'>
          <h5 className="card-title"> {superhero} </h5>
          <p className="card-text"> {publisher} </p>
          <div className='d-flex flex-row '>
            <p className="card-text p-2"> {alter_ego} </p>
            <p className="card-text p-2">
              {
                ( alter_ego !== characters ) && 
                  <small className='text-muted'> {characters} </small>
              }
            </p>
          </div>
          <p className="card-text p-2"> {first_appearance} </p>
        </div>
        
        <Link to={`/hero/${id}`}>
          <button
            className='btn btn-dark w-100'
            type='button'
          >
            Mas informacion <img src={infoImagePath} alt='info' />
          </button>
        </Link>
      </div>
    </div>
  )
};
