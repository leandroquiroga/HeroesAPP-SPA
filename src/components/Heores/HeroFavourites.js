import React from 'react'
import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';



export const HeroFavourites = (props) => {
  const { data, pathCloseSVG, dispatchHero } = props;

  const { id, superhero, publisher, characters, first_appearance } = data;

  const handleDeleteHero = ((heroID) => {
  
    
    dispatchHero({
      type: 'deleteHero',
      payload: heroID
    });

    
    (function notify() {
      toast.error('Se ha eliminado un heroe', {
        theme: 'colored',
        position: "top-right",
        autoClose: 2000,
        hideProgressBar: false,
        closeOnClick: true,
        pauseOnHover: true,
        draggable: true,
        progress: undefined,
      });
    })();
  });

  return (
    <div
      className='col animate__animated animate__fadeIn px-2'
    >
      <li
        className='card'
      >
        <div className='d-flex justify-content-end align-items-center'>
          <button
            className='main-container--button'
          >
            <img
              src={pathCloseSVG}
              alt='delete'
              className='main-container--button-close'
              onClick={() => handleDeleteHero(id)}
            />
          </button>
          <ToastContainer />
        </div>
        <img 
          src={`/assets/img/${id}.jpg`}
          alt={superhero}
          className='card-img-top img-thumbnail'
        />
        
        <div className='card-body d-flex flex-column justify-content-center align-items-center'>
          <h5 className="card-title"> {superhero} </h5>
          <p className="card-text"> {publisher} </p>
          <div className='d-flex flex-row '>
            <p className="card-text p-2">
                  <small className='text-muted'> {characters} </small>
            </p>
          </div>
          <p className="card-text p-2"> {first_appearance} </p>
        </div>
      </li>
    </div>
  )
}
