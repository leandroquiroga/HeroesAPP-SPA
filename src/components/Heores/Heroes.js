import React, { useMemo } from 'react';
import { useParams, useNavigate, Navigate } from 'react-router-dom';
import { getHeroById } from './../../helpers/index';
import { ToastContainer, toast } from 'react-toastify';

import 'react-toastify/dist/ReactToastify.css';
export const Heroes = () => {
  // const [msgAdd , setMsgAdd] = useState(false)  
  const { heroID } = useParams();
  const navigate = useNavigate();
  const hero = useMemo(() => getHeroById(heroID), [heroID] ) 

  const { id, superhero, alter_ego, characters, first_appearance, publisher } = hero; 

  if (!hero) {
    return <Navigate to='/' />
  };

  // Regresa al url anterior de la pila del stack
  const handleBack = () => navigate(-1);


  const notify = () => {
    toast.success('Nuevo heroe agregado', {
      theme: 'colored',
      position: "bottom-right",
      autoClose: 2000,
      hideProgressBar: false,
      closeOnClick: true,
      pauseOnHover: true,
      draggable: true,
      progress: undefined,
      });
  }
  

  const imagePath = `/assets/img/${id}.jpg`
  const startImagePath = `/assets/svg/start.svg`
  const backImagePath = `/assets/svg/back.svg`
  return (
    <div className='container animate__animated animate__fadeIn '>
      <div className='d-flex flex-column mt-3 flex-md-row '>
        <div className='col-12 col-md-4 '>
          <img
            src={imagePath}
            alt={superhero}
            className='img-fluid w-100'
          >
          </img>
        </div>
        <div className='col-12 col-md-8 p-5'>
            <h3>{superhero}</h3>
            <ul className='list-group list-group-flush'>
              <li className='list-group-item'>
                <b>Alter ego: </b>
                  {alter_ego}
              </li>
              <li className='list-group-item'>
                <b>Publisher: </b>
                  {publisher}
              </li>
              <li className='list-group-item'>
                <b>First Appearance: </b>
                  {first_appearance}
              </li>
            </ul>

            <h5 className='mt-3 mx-3'> Characters </h5>
            <p className='mx-3'> {characters} </p>

          <div className='d-flex flex-row '>
              <button
                className='btn btn-danger m-2'
                onClick={handleBack}
              >
                Regresar <img src={backImagePath} alt='favorite' />
              </button>
              <div className='d-flex flex-column flex-md-row align-items-center'>
                <button
                  className='btn btn-danger m-2'
                  onClick={notify}>
                  AÃ±adir a <img src={startImagePath} alt='favorite' />
              </button>
              <ToastContainer />
            </div>  
          </div>

        </div>
      </div>
    </div>
  );
};
