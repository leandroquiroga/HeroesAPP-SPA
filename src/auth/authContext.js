import { createContext, useEffect, useReducer, useState } from "react";
import { signInWithEmailAndPassword, signOut ,createUserWithEmailAndPassword } from 'firebase/auth';
import { authReducer } from "./authReducer";
import { auth } from './../firebase';
import { heroReducer } from "../helpers";

// Retorna el usuario si lo encuentra en localStorage si no retorna el logged en false
const init = () => JSON.parse(localStorage.getItem('user')) || {logged: false}

// Retorna los hero si lo encuntra en localStorage si no retorna un array 
const initHero = () => JSON.parse(localStorage.getItem('hero')) || [];

export const AuthContext = createContext();

export const AuthProvaider = (props) => {
  const [user, dispatch] = useReducer(authReducer, {}, init);
  const [currentUser, setCurrentUser] = useState({});
  const [hero, dispatchHero] = useReducer(heroReducer, [], initHero);

  // Cuando inicia la aplicacion va a estar observando si hay un cambio
  // de usuario.
  useEffect(() => {
    auth.onAuthStateChanged((user) => setCurrentUser(user))
  }, []);

  // Esta pendiente al user para ir almancendolo en el localStorage
  useEffect(() => {
    localStorage.setItem('user', JSON.stringify(user))
    return () => {
      if (!user) return;
    }
  }, [user]);

  // Esta pendiente al hero, cuando detecta un cambio lo almacena en el localStorage
  useEffect(() => {
    localStorage.setItem('hero', JSON.stringify(hero));
    return () => {
      if (!hero) return;
    }
  }, [hero])
  


  const singUp = (email, password) => createUserWithEmailAndPassword(auth, email, password);

  const logIn = (email, password) => signInWithEmailAndPassword(auth, email, password);
  
  const logOut = (auth) => signOut(auth);

  const value = {
    user,
    dispatch,
    singUp,
    logIn,
    logOut,
    currentUser,
    dispatchHero,
    hero,
  }
  return (
    <AuthContext.Provider value={value}>
      {props.children}
    </AuthContext.Provider>
  )
}