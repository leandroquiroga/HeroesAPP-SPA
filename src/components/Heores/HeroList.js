import React, { useMemo } from 'react';
import { getHeroByPublisher } from '../../helpers';
import { HeroCard } from './HeroCard';


export const HeroList = ({publisher}) => {
  
  const heroes = useMemo(() => getHeroByPublisher(publisher), [publisher]);

  return (
    <div className=' animate__animated animate__fadeIn row justify-content-center row-cols-1 row-cols-md-2  row-cols-xxl-3 g-2'>
      
      {
        heroes.map(hero => ( 
          <HeroCard
            key={hero.id}
            {...hero}   
          />
        ))
      }
      
    </div>
  );
};