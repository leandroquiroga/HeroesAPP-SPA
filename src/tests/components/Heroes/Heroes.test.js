
import { Heroes } from './../../../components/Heores/Heroes';
import { mount } from 'enzyme';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { AuthContext } from '../../../auth/authContext';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}))

describe('Testing on <Heroes />', () => {
  const heroReducer = {
    dispatchHero: jest.fn(),
    hero: [
      {
        'id': 'dc-batman',
        'superhero': 'Batman',
        'publisher': 'DC Comics',
        'alter_ego': 'Bruce Wayne',
        'first_appearance': 'Detective Comics #27',
        'characters': 'Bruce Wayne'
      },
    ],
  }
  test('This test must should display the component corretcly', () => {

    const wapper = mount(
      <AuthContext.Provider value={heroReducer}>

        <MemoryRouter initialEntries={['/hero/dc-batman']}>
          <Routes>
            <Route path='/hero/:heroID' element={<Heroes />} />
            <Route path='/' element={<h1>No hay heroes</h1>} />
          </Routes>
        </MemoryRouter>
      
      </AuthContext.Provider>
    );
    
    expect(wapper).toMatchSnapshot();
    expect(wapper.find('h3').text().trim()).toBe('Batman')
  });

  test('This test should display an message "No hay heroes" if not exist an hero', () => {
    
    const wapper = mount(
      <AuthContext.Provider value={heroReducer}>

        <MemoryRouter initialEntries={['/']}>
          <Routes>
            <Route path='/hero' element={<Heroes />} />
            <Route path='/' element={<h1>No hay heroes</h1>} />
          </Routes>
        </MemoryRouter>
      
      </AuthContext.Provider>
    );

    expect(wapper.find('h1').text().trim()).toBe('No hay heroes');
  });

  test('This test should return to the previos screen', () => {
    const wapper = mount(
      <AuthContext.Provider value={heroReducer}>

        <MemoryRouter initialEntries={['/hero/dc-batman']}>
          <Routes>
            <Route path='/hero/:heroID' element={<Heroes />} />
          </Routes>
        </MemoryRouter>
      
      </AuthContext.Provider>
    );

    wapper.find('button').at(0).prop('onClick')();

    expect(mockNavigate).toHaveBeenCalledWith(-1);

  })
})