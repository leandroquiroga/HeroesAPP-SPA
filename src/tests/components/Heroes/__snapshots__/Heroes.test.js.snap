// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing on <Heroes /> This test must should display the component corretcly 1`] = `
<div
  className="container animate__animated animate__fadeIn "
>
  <div
    className="d-flex flex-column mt-3 flex-md-row "
  >
    <div
      className="col-12 col-md-4 "
    >
      <img
        alt="Batman"
        className="img-fluid w-100"
        src="/assets/img/dc-batman.jpg"
      />
    </div>
    <div
      className="col-12 col-md-8 p-5"
    >
      <h3>
        Batman
      </h3>
      <ul
        className="list-group list-group-flush"
      >
        <li
          className="list-group-item"
        >
          <b>
            Alter ego: 
          </b>
          Bruce Wayne
        </li>
        <li
          className="list-group-item"
        >
          <b>
            Publisher: 
          </b>
          DC Comics
        </li>
        <li
          className="list-group-item"
        >
          <b>
            First Appearance: 
          </b>
          Detective Comics #27
        </li>
      </ul>
      <h5
        className="mt-3 mx-3"
      >
         Characters 
      </h5>
      <p
        className="mx-3"
      >
         
        Bruce Wayne
         
      </p>
      <div
        className="d-flex flex-row "
      >
        <button
          className="btn btn-danger m-2"
          onClick={[Function]}
        >
          Regresar 
          <img
            alt="favorite"
            src="/assets/svg/back.svg"
          />
        </button>
        <div
          className="d-flex flex-column flex-md-row align-items-center"
        >
          <button
            className="btn btn-danger m-2"
            onClick={[Function]}
          >
            AÃ±adir a 
            <img
              alt="favorite"
              src="/assets/svg/start.svg"
            />
          </button>
          <div
            className="Toastify"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
